mbedtls/src/dsa/mod.rs

```diff
@@ -0,0 +1,545 @@
+        let q = Mpi::from_binary(&q.to_bytes_be()).expect("Success");
+        let g = Mpi::from_binary(&g.to_bytes_be()).expect("Success");
+        let y = Mpi::from_binary(&y.to_bytes_be()).expect("Success");
+
+        let params = DsaParams::from_params(p, q, g)?;
+
+        DsaPublicKey::from_params(params, y)
+    }
+
+    pub fn to_der(&self) -> Result<Vec<u8>> {
+        let p = BigUint::from_bytes_be(&self.params.p.to_binary()?);
+        let q = BigUint::from_bytes_be(&self.params.q.to_binary()?);
+        let g = BigUint::from_bytes_be(&self.params.g.to_binary()?);
+        let y = BigUint::from_bytes_be(&self.y.to_binary()?);
+
+        let y_as_int = yasna::construct_der(|w| {
+            w.write_biguint(&y)
+        });
+
+        let der = yasna::construct_der(|w| {
+            w.write_sequence(|w| {
+                w.next().write_sequence(|w| {
+                    w.next().write_oid(&ObjectIdentifier::from_slice(DSA_OBJECT_IDENTIFIER));
+                    w.next().write_sequence(|w| {
+                        w.next().write_biguint(&p);
+                        w.next().write_biguint(&q);
+                        w.next().write_biguint(&g);
+                    });
+                });
+                w.next().write_bitvec(&BitVec::from_bytes(&y_as_int));
+            })
+        });
+
+        Ok(der)
+    }
+
+    pub fn verify(&self, signature: &[u8], pre_hashed_message: &[u8]) -> Result<bool> {
+        if signature.len() % 2 == 1 {
+            return Ok(false);
+        }
+
+        let p = &self.params.p;
+        let q = &self.params.q;
+
+        let half = signature.len() / 2;
+        let r = Mpi::from_binary(&signature[..half])?;
+        let s = Mpi::from_binary(&signature[half..])?;
+
+        let zero = Mpi::new(0)?;
+
+        if r == zero || s == zero {
+            return Ok(false);
+        }
+
+        if &r >= q || &s >= q {
+            return Ok(false);
+        }
+
+        let m = reduce_mod_q(pre_hashed_message, q)?;
+
+        let s_inv = s.modinv(q)?;
+
+        let sr = (&s_inv * &r)?.modulo(q)?;
+        let sm = (&s_inv * &m)?.modulo(q)?;
+
+        // Compute (g^sm * y^sr) mod p
+        // mbedtls doesn't support multi-exponentiation
+        let gsm = self.params.g.mod_exp(&sm, p)?;
+        let ysr = self.y.mod_exp(&sr, p)?;
+        let gsm_ysr = (&gsm * &ysr)?.modulo(p)?;
+
+        Ok(gsm_ysr.modulo(q)? == r)
+    }
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct DsaPrivateKey {
+    params: DsaParams,
+    x: Mpi,
+}
+
+fn sample_secret_value<F: Random>(upper_bound: &Mpi, rng: &mut F) -> Result<Mpi> {
+    /*
+    See FIPS 186-4 Appendix B.2.1
+     */
+    let bits = upper_bound.bit_length()?;
+    let mut rnd_buf = vec![0u8; (bits + 7 + 64) / 8];
+    rng.random(&mut rnd_buf)?;
+    let c = Mpi::from_binary(&rnd_buf)?;
+    let mut c = c.modulo(&(upper_bound - 1)?)?;
+    c += 1;
+    Ok(c)
+}
+
+fn encode_dsa_signature(q_bits: usize, r: &Mpi, s: &Mpi) -> Result<Vec<u8>> {
```

Comment:
On second thought, for ECDSA, we use DER-encoded signatures. Is this specified the same way for DSA? We should probably use that.

Reply:
Yeah for DSA and ECDSA both methods are about equally common depending on the standard. DSA in X.509 uses DER IIRC.

Reply:
Done